WIP Brave 4.1 APIs

llinder opened the pull request:
Very very rough and untested attempt at using the Brave 4.1 APIs.

Next steps is to actually test this out and add abstractions where necessary.

llinder committed:
WIP Brave 4.1 APIs

codefromthecrypt commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -81,18 +75,16 @@ public void close() {
-                        .handle(responseAdapterFactory.createAdapter(receivedResponseResult.getValue()))
-                );
+        final ExecuteWithSpan execution = new ExecuteWithSpan(tracer);
+        return delegate.request(uri, execution.execute(action))
```
it might be useful to eagerly provision a span, then wrap delegate.request with `try(SpanInScope ws = tracer.withSpanInScope(span))` as that would allow anything invoking downstream see it (potentially helpful if SLF4J MDC is in use). You can lazy start that span in the execution thing..

codefromthecrypt replied:
Key part is brave 4 spans are started independent of their creation exactly for this reason..

tracer.nextSpan(); for example, will create a span.. you can set that in scope independent of its timing, and setting it will allow others to read it (ex to add tags or read trace IDs).

codefromthecrypt commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -105,14 +97,64 @@ public void close() {
+    private void tracedResponse(Result<?> result, StatusCode statusCode, @Nullable Span span) {
+        if (span != null) {
+            if (result.isError()) {
+                span.tag(Constants.ERROR, result.getThrowable().getMessage());
```
potential NPE during the tagging.. 

I usually use something like..
```java
String message = throwable.getMessage();
if (message == null) message = throwable.getClass().getSimpleName();
```

codefromthecrypt commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -105,14 +97,64 @@ public void close() {
+
+        Action<? super RequestSpec> execute(Action<? super RequestSpec> action) {
+            return (request -> {
+                Span parent = tracer.currentSpan();
```
tracer.nextSpan() is a shortcut for this dance

llinder replied:
Thanks for the suggestion.  It wasn't clear to me if this would start a new trace if one wasn't already in progress.  I will switch to `tracer.nextSpan()` instead.

codefromthecrypt reviewed (commented):
cool start! will be watching

hyleung commented:
@llinder wow! I started working on the same thing last night - but clearly you're a lot further along than me. 😄

hyleung commented:
@llinder looks like there's a bunch of stuff that's going to be outright removed as part of this change. If that's the direction we're going in, then I suppose we might be looking at a major version bump (since there would be breaking changes)?

llinder commented:
@hyleung glad to hear you are looking into this as well!  Right now I'm just considering this a spike.  If the directions seems good then we can add parts back in that make sense for Brave 4.

My hope is that the changes will have minimal to no effect for most folks.  For anyone interacting with the Brave 3 APIs and wishing to upgrade there will definitely be some breaking changes.  We could always go a different direction and keep both Brave 3 and Brave 4 APIs in the project but that feels overly complicated.  My preference is to just use Brave 4 and do a version bump.  Then again I don't want to make decisions for other folks so hopefully we will get some feedback.

Also I don't want to feel like I'm taking over so feel free to steer this as you see fit and I will follow your lead :)

hyleung commented on src/main/java/ratpack/zipkin/ServerTracingModule.java:
```diff
@@ -15,107 +15,60 @@
-    }
-    if (config.sampler != null) {
-      braveBuilder.traceSampler(config.sampler);
-    }
```
I couldn't resist building your branch locally and giving it a spin - got an NPE due because I hadn't included a `Sampler` in my config. The way it currently is, it's optional - and I kinda like having a "batteries included" set of defaults so maybe that's something we can consider...

llinder replied:
Thanks for flagging this.  Definitely should include defaults here!

hyleung commented:
@llinder at the moment (we'll see how long it lasts) I'm trying to take an incremental approach but yeah, I have a hunch that over time a version bump with breaking changes might be the easiest. If it comes to that, them maybe we just get one more release out to get things synced up to the last 3.x Brave release (plus any PRs that might be relevant) leave it at that...

hyleung commented on src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java:
```diff
@@ -0,0 +1,41 @@
+  private final MDCProxy mdc;
+
+  private final TraceContext.Extractor<MDCProxy> extractor = Propagation.B3_STRING.extractor(MDCProxy::get);
+  private final TraceContext.Injector<MDCProxy> injector = Propagation.B3_STRING.injector(MDCProxy::put);
```
What sorcery is this? 😁

codefromthecrypt replied:
clever.. I never thought of doing it this way!

So, this is literally using the slf4j as the thread local storage for zipkin tracing. brilliant @llinder

codefromthecrypt replied:
impact is that in your logging configuration, you would use the keys named same as headers (ex X-B3-TraceId). Some currently use lowercamel variants like traceId. changing the key format to something different than the header names might be requested in the future, but this is a reasonable default IMHO, and so little code compared to how I was thinking.

llinder replied:
I think this can all be removed.  It was basically an attempt at using the MDC context for span context propagation.  It isn't really working as intended.  Will switch to the Ratpack MutableRegistry approach instead.

hyleung replied:
~I suppose you'd also be able to correlate your logs with Zipkin traces for free too!~ Doh! It could have been so beautiful!

codefromthecrypt replied:
:)

llinder replied:
Log correlation would be a good reason to keep it :)  Easy to add back I suppose, still I seem to have a bug in here or Ratpack has a bug in their MDCInterceptor because it didn't seem to work.  I was probably just doing something stupid though ;)

codefromthecrypt commented on src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java:
```diff
@@ -0,0 +1,41 @@
+    return () -> injector.inject(previous, mdc);
+  }
+
+  private static final class MDCProxy {
```
suppose proxy is needed as you can't get method reference to a static method?

hyleung replied:
This proxy thing is probably from my obsession about being able to test everything (iirc) - it's just some indirection to provide a seam for testing.

llinder replied:
Yeah thats exactly how I ended up with this :)  I don't think the MDC approach is the ideal approach with Ratpack.  It might be useful to put the trace context in MDC for other reasons but I think we can do without it for now.

codefromthecrypt commented:
>
> I think this can all be removed. It was basically an attempt at using the
> MDC context for span context propagation. It isn't really working as
> intended. Will switch to the Ratpack MutableRegistry approach instead.
>
reminds me to make a base test class for CurrentTraceContext impl.  first
guess is that MDC would need some sort of dance to restore properly on
close.. anyway nice try very interesting!

https://github.com/openzipkin/brave/blob/master/brave/src/test/java/brave/features/log4j2_context/Log4JThreadContextTest.java

^^ has a dispatch approach which could still do log correlation, but uses a
more explicit means of attaching scope

codefromthecrypt commented:
>
> This proxy thing is probably from my obsession about being able to test
> everything (iirc) - it's just some indirection to provide a seam for
> testing.
>
heh powermockito is a sledge hammer I sometimes use (but it can become a
cure worse than the disease)

llinder commented:
By the way, do people prefer I keep commit logs for all changes or is squashing ok?  I tend to do the later but having the history to look back on might help.

hyleung commented:
@llinder ¯\\_(ツ)_/¯  generally i'm pretty cool with whatever folks want to do on their own forks (tho' I do prefer a squash before the final merge). 

Having said that, if you're up for it, I'd love to be able to collaborate on this PR (even if it's just pitching in on tidying things up, writing tests, etc.) - in which case, obviously squashes on WIP is a a bit problematic.

llinder committed:
Trace context propagation for ParallelBatch

llinder commented:
@hyleung cool I will avoid squashes to make collaboration easier.

llinder commented on src/main/java/ratpack/zipkin/TracedParallelBatch.java:
```diff
@@ -0,0 +1,34 @@
+   *
+   * Example usage:
+   *
+   * TracedParallelBatch
```
Pulled in the ParallelBatch work that @edwardsoo-ss was working on.  Essentially the same thing except I made some helper methods to abstract away the `.execInit` stuff.

llinder commented on src/main/java/ratpack/zipkin/TracedParallelBatch.java:
```diff
@@ -0,0 +1,34 @@
+import ratpack.exec.util.ParallelBatch;
+import ratpack.zipkin.internal.RatpackCurrentTraceContext.TraceContextTypeValue;
+
+public class TracedParallelBatch {
```
Pulled in the ParallelBatch work that @edwardsoo-ss was working on. Essentially the same thing except I made some helper methods to abstract away the .execInit stuff.

edwardsoo-ss replied:
:+1: thanks

edwardsoo-ss commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -105,14 +109,63 @@ public void close() {
-            requestInterceptor
-                    .handle(requestAdapterFactory.createAdaptor(requestSpec.method(capturedMethod), capturedMethod.getName()));
+            span.name(spanNameProvider.getName(new DefaultRequestSpecNameAdapter(captor)));
+            span.tag(TraceKeys.HTTP_URL, captor.getUri().toString());
```
maybe use `ratpack.zipkin.internal.ZipkinHttpClientImpl.DefaultRequestSpecNameAdapter#getUri` just to be more consistent?

edwardsoo-ss commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -105,14 +109,63 @@ public void close() {
     }
 
-    private Action<? super RequestSpec> tracedRequestAction(final Action<? super RequestSpec> action) {
-        return action.append(requestSpec -> {
```
I didn't test this out, but I don't think the original action will be applied to the actual `RequestSpec` without this append/prepend

edwardsoo-ss replied:
alternatively we can make `ratpack.zipkin.internal.MethodCapturingRequestSpec` modifies the actual `ratpack.http.client.RequestSpec` in its spec methods

edwardsoo-ss commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -81,18 +79,24 @@ public void close() {
-                );
+        final Span span = tracer.nextSpan();
+        try(Tracer.SpanInScope ws = tracer.withSpanInScope(span)) {
+            span.start();
```
this is called in `actionWithSpan`, probably not needed?

edwardsoo-ss commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -81,18 +79,24 @@ public void close() {
-                );
+        final Span span = tracer.nextSpan();
+        try(Tracer.SpanInScope ws = tracer.withSpanInScope(span)) {
+            span.start();
```
ditto

edwardsoo-ss commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -105,14 +109,63 @@ public void close() {
+    private Action<? super RequestSpec> actionWithSpan(final Action<? super RequestSpec> action, final Span span) {
+        return (request -> {
+            MethodCapturingRequestSpec captor = new MethodCapturingRequestSpec(request);
+            span.start();
```
I think we should also annotate "cs" here

edwardsoo-ss commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -105,14 +109,63 @@ public void close() {
+                span.tag(Constants.ERROR, result.getThrowable().getClass().getSimpleName());
+            }
+        } else {
+            int status = statusCode.getStatusCode();
```
annotate "cr" around here?

llinder replied:
cs/cr annotations are handled by the tracer.  Thanks for bringing it up though because I did forget to set the span kind to client here.

edwardsoo-ss replied:
ah I see `brave.Span#kind` :+1:

edwardsoo-ss commented on src/main/java/ratpack/zipkin/internal/DefaultServerTracingHandler.java:
```diff
@@ -1,44 +1,90 @@
+    Request request = ctx.getRequest();
+    span.tag(TraceKeys.HTTP_URL, request.getUri());
+
+    ctx.getResponse().beforeSend(response -> {
```
missing "ss" here and "sr" below?

edwardsoo-ss commented:
I will try it out locally, looking good so far!

llinder committed:
Trace context and http client fixes plus tests

llinder committed:
Set correct span kind for client calls

llinder commented:
Thanks for the comments and catching my mistakes @edwardsoo-ss!

llinder committed:
Fix missing ss/sr annotations

hyleung commented:
@llinder Dude you're on 🔥! 

I was thinking about this a bit and it looks like you're on the path to pretty much on the path to doing away with `brave-http` altogether - i.e. getting down to the nitty gritty of annotating those spans yourself. What do you figure is the benefit of this approach over using `brave-http`'s APIs (all the adapters and whatnot)?

codefromthecrypt commented:
brave-http will mostly go away once more generic Client/ServerHandler
support is in
https://github.com/openzipkin/brave/pull/350/files#diff-e0f8554fb22e6ddd2bc65c6ea53965e2

If it remains, it would only include helpers related to http tags I think.

On Thu, Apr 6, 2017 at 11:18 AM, Ho Yan Leung <notifications@github.com>
wrote:

> @llinder <https://github.com/llinder> Dude you're on 🔥!
>
> I was thinking about this a bit and it looks like you're on the path to
> pretty much on the path to doing away with brave-http altogether - i.e.
> getting down to the nitty gritty of annotating those spans yourself. What
> do you figure is the benefit of this approach over using brave-http's
> APIs (all the adapters and whatnot)?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyleung/ratpack-zipkin/pull/24#issuecomment-292057862>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD61_U19sHye9oN1J_KxPpVzGOWPfcWks5rtFlpgaJpZM4MyV-_>
> .
>

hyleung commented:
@adriancole ah, ic...so it would be a case of implementing (in this case) RatPack specific subclasses of these handlers? In any case, the logic around dealing with starting/finishing spans, making sure they're correctly annotated, etc. - that would reside in the base classes that are provided by `Brave` (e.g. `ServerHandler.handleRecieve`, `ServerHandler.handleSend`, right?

codefromthecrypt commented:
>
> @adriancole <https://github.com/adriancole> ah, ic...so it would be a
> case of implementing (in this case) RatPack specific subclasses of these
> handlers? In any case, the logic around dealing with starting/finishing
> spans, making sure they're correctly annotated, etc. - that would reside in
> the base classes that are provided by Brave (e.g.
> ServerHandler.handleRecieve, ServerHandler.handleSend, right?
>
The end goals of that PR are the following... not all would apply here

* be able to switch out B3 if needed (ex for internal things or another
format) that's the Propagation type
* be able to switch out the context implementation (even if ratpack
decorates, would be nice for folks to be able to weave in SLF4J etc)
* separate tagging implementation from tagging policy (ex be able to reduce
the tags emitted without knowing how they are created.. ex turn off
http.url or add a custom tag)
* have a single test framework that makes sure things mostly work
regardless of how they are implemented

So, once this is in, we'd have tests to say if you could add response.size
tag, for example, or to disable client address parsing regardless of
library.
I'm kindof meh about sharing base classes for this as some of the code
might be doable in another way.. sharing some config classes (especially
saying which propagation to use and what tags to record) might be helpful
regardless as that would allow someone a single switch.

what do you think?

llinder commented:
Thanks @hyleung but to his credit @adriancole is the one on 🔥 I'm just attempting to keep up with his breakneck pace :)   PR 350 looks like a great direction and I could see a lot of value using the pluggable span name, tags and address parsers here as well.

I mostly left `brave-http` out of the picture in this PR to get a better feeling for the core Brave 4 APIs without abstractions.  In general I found it way easier to use as apposed to Brave 3 and that the extra abstractions didn't provide as much value.  On the contrary, having a consistent way to handle span name, tag and address parsing seems very desirable, it just wasn't my initial focus.  Now that something has mostly been proved out with other frameworks I think it totally makes sense to apply them here as well.

hyleung commented:
@adriancole as always, thanks for the additional context :)

Re. `brave-http`, base classes, etc., for me the primary thing is being able to verify that a particular implementation has the correct behaviour. Especially at the beginning (hell, it's probably still the case now) when I wasn't super familiar with what needed to happen at a lower level, being able to implement a few adapters and interfaces and have it just plug in was handy. So having a way of verifying an implementation, whether it be that "single test framework" or otherwise would definitely be helpful.

hyleung committed:
Refactor/add tests

hyleung commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -105,14 +106,64 @@ public void close() {
+    }
+
+    private void resultWithSpan(Result<?> result, StatusCode statusCode, Span span) {
+        if (result.isError()) {
```
@llinder I've been trying to write some tests for this class and can't for the life of me figure out how to get this branch to execute. Under what circumstances would we expect `result.isError()` to evaluate to true?

hyleung committed:
PR feedback

llinder committed:
Merge pull request #1 from hyleung/add_tests

Refactor/add tests

llinder committed:
Update to brave 4.2

hyleung commented:
@llinder could I trouble you with one small thing on this PR before it's merged - can we split the `ParallelBatch` [stuff](https://github.com/hyleung/ratpack-zipkin/pull/24/commits/616eb34f80ed7d000ab75d5a7fb03613873fe8fc) into a separate PR? I had a question from https://github.com/hyleung/ratpack-zipkin/pull/23, plus I'd prefer not to have it mixed in with the Brave upgrade.

Other than that, let's get this merged in soon. What I'll do is probably one more 1.x release (for https://github.com/hyleung/ratpack-zipkin/pull/21), then bump it to `2.0.0-SNAPSHOT` and merge this in.

Does that sound reasonable?

[Update: I've gone ahead and done the bump to 2.0.0]

llinder commented:
Sounds good.  I will split the parallel batch stuff off.  I would like to get test coverage up a little more so I will try to tackle some of that this week as well.  Or we can address test coverage post merge, either way works for me.

hyleung commented on src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java:
```diff
@@ -105,14 +106,64 @@ public void close() {
+    }
+
+    void resultWithSpan(Result<?> result, StatusCode statusCode, Span span) {
+        if (result.isError()) {
```
@llinder I wasn't able to write a test to exercise this code path - maybe I had the setup wrong?

hyleung commented:
@llinder I did have one [bit of the code](https://github.com/hyleung/ratpack-zipkin/pull/24/files#r112861075) that I had trouble writing a test for.

Other than that, I'm ok with addressing coverage after merge (I'm happy to help out with that).  I think the JaCoCo doesn't deal with lambdas very well so the coverage report is a bit misleading sometimes.

llinder committed:
Expose Tracing component and inject client tracing headers

llinder commented:
@hyleung I removed the parallel batch stuff for now.  Also updated the dependency injection to use the Tracing component as suggested by @adriancole.  Lastly I updated the Ratpack request so that the tracing context headers are injected into the request.

Hopefully this is in a fairly good state to merge.  I will keep tracing Brave #350 work and incorporate new additions as they are released.

codefromthecrypt commented:
Hopefully this is in a good state to.. MERGED

Nice work and thanks for being the earliest into the new apis. I know it is
especially tricky with async.

hyleung commented:
@adriancole haha…to be clear, there are still a few bits an pieces (e.g. features that were removed) that I'd like to add back before we do a release - but we figured it would be better to make some forward progress.

@llinder I'm gonna add a couple of issues for the missing features (probably later today). If there's anything that you think needs to be included for an initial 2x. release feel free to add to the list (even if it's just a placeholder so we remember to discuss it) 😄
